File: leetcode_6/src/leetcode_6/max_level_sum.rs
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    #[allow(dead_code)]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}
struct Solution;
use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    #[allow(dead_code)]
    pub fn max_level_sum(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let mut cur = vec![root];
        let mut next = vec![];
        let mut max_level = 0;
        let mut cur_max = i32::MIN;
        let mut level = 0;
        while !cur.is_empty() {
            level += 1;
            let sum = cur
                .into_iter()
                .map(|y| {
                    let unwrap = y.unwrap();
                    let mut m = unwrap.borrow_mut();
                    let l = m.left.take();
                    let r = m.right.take();
                    if l.is_some() {
                        next.push(l);
                    }
                    if r.is_some() {
                        next.push(r);
                    }
                    m.val
                })
                .sum::<i32>();
            if sum > cur_max {
                cur_max = sum;
                max_level = level;
            }
            cur = next;
            next = vec![];
        }
        max_level
    }
}

File: leetcode_cpp/Leetcode75/BinaryTree/maxLevelSum.cpp
#include <bits/stdc++.h>
using namespace std;
auto init = []() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    return 'c';
}();

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode():
      val(0),
      left(nullptr),
      right(nullptr) {}

    explicit TreeNode(int x):
      val(x),
      left(nullptr),
      right(nullptr) {}

    TreeNode(int x, TreeNode *left, TreeNode *right):
      val(x),
      left(left),
      right(right) {}
};

class Solution {
  public:
    static int maxLevelSum(TreeNode *root) {
        if (!root) {
            return 0;
        }
        std::vector<TreeNode *> v{root};
        std::vector<TreeNode *> tmp;
        int max = std::numeric_limits<int>::min();
        int min_level = 0;
        int level = 1;
        while (!v.empty()) {
            auto sum = std::accumulate(v.begin(), v.end(), 0, [](int sum, TreeNode *node) { return sum + node->val; });
            if (max < sum) {
                max = sum;
                min_level = level;
            }
            for (auto node : v) {
                if (node->left) {
                    tmp.push_back(node->left);
                }
                if (node->right) {
                    tmp.push_back(node->right);
                }
            }

            level++;
            v.swap(tmp);
            tmp.clear();
        }
        return min_level;
    }
};